include/yaml-cpp/node/convert.h

```diff
@@ -161,109 +170,222 @@ struct convert<bool> {
+    for (const auto& item : node)
+      sequence.push_back(item.template as<typename T::value_type>());
+    return true;
+  }
+};
+
+// Helper class for converting something which is "map-like" to and from a Node.
+template <class T, class Key = typename T::key_type,
+          class Value = typename T::mapped_type>
+struct convert_as_map {
+  static Node encode(const T& map) {
     Node node(NodeType::Map);
-    for (typename std::map<K, V>::const_iterator it = rhs.begin();
-         it != rhs.end(); ++it)
-      node.force_insert(it->first, it->second);
+    for (const auto& kv : map)
+      node.force_insert(kv.first, kv.second);
     return node;
   }
 
-  static bool decode(const Node& node, std::map<K, V>& rhs) {
+  static bool decode(const Node& node, T& map) {
     if (!node.IsMap())
       return false;
+    map.clear();
+    for (const auto& kv : node) {
+// if we're dealing with GCC (note: clang also defines __GNUC__)
+#if defined(__GNUC__)
+#if __GNUC__ < 4
+      // gcc version < 4
+      map.insert(std::make_pair(kv.first.template as<Key>(),
+                                kv.second.template as<Value>()));
+#elif __GNUC__ == 4 && __GNUC_MINOR__ < 8
+      // 4.0 <= gcc version < 4.8
+      map.insert(std::make_pair(kv.first.as<Key>(), kv.second.as<Value>()));
+#else
+      // 4.8 <= gcc version
+      map.emplace(kv.first.as<Key>(), kv.second.as<Value>());
+#endif  // __GNUC__ < 4
+#else
+      // for anything not GCC or clang...
+      // probably some more #ifdef guards are needed for MSVC.
+      map.emplace(kv.first.as<Key>(), kv.second.as<Value>());
+#endif  // defined(__GNUC__)
+    }
+    return true;
+  }
+};
+
+// Helper class for converting something which is "set-like" to and from a Node.
+// A set is realized as a yaml sequence.
+template <class T>
+struct convert_as_set {
+  static Node encode(const T& set) {
+    Node node(NodeType::Sequence);
+    for (const auto& item : set)
+      node.push_back(item);
+    return node;
+  }
+
+  static bool decode(const Node& node, T& set) {
+    if (!node.IsSequence())
+      return false;
+
+    set.clear();
+    for (const auto& item : node) {
 
-    rhs.clear();
-    for (const_iterator it = node.begin(); it != node.end(); ++it)
 #if defined(__GNUC__) && __GNUC__ < 4
       // workaround for GCC 3:
-      rhs[it->first.template as<K>()] = it->second.template as<V>();
-#else
-      rhs[it->first.as<K>()] = it->second.as<V>();
+      set.insert(item.template as<typename T::value_type>());
+#else  // __GNUC__ is not defined or __GNUC__ >= 4
+      set.insert(item.as<typename T::value_type>());
 #endif
+    }
     return true;
   }
 };
 
-// std::vector
-template <typename T>
-struct convert<std::vector<T>> {
-  static Node encode(const std::vector<T>& rhs) {
+// Helper class for converting something which is "multimap-like" to and from a
+// Node. A multimap is realized as a yaml sequence of key-value pairs.
+template <class T>
+struct convert_as_multimap {
```

Comment:
I think we should nix multimap; it's not in the spec, and there are various ways you could do it.

In particular, a sequence of key-value pairs is an ordered map, according to the spec:

http://yaml.org/type/omap.html